pkg/proxy/config/api.go

```diff
@@ -21,223 +21,41 @@ import (
-}
-
-// SourceAPI implements a configuration source for services and endpoints that
-// uses the client watch API to efficiently detect changes.
-type SourceAPI struct {
-	s servicesReflector
-	e endpointsReflector
-}
-
-type servicesReflector struct {
-	watcher           ServicesWatcher
-	services          chan<- ServiceUpdate
-	resourceVersion   string
-	waitDuration      time.Duration
-	reconnectDuration time.Duration
-}
-
-type endpointsReflector struct {
-	watcher           EndpointsWatcher
-	endpoints         chan<- EndpointsUpdate
-	resourceVersion   string
-	waitDuration      time.Duration
-	reconnectDuration time.Duration
-}
+// NewSourceAPIserver creates config source that watches for changes to the services and endpoints.
+func NewSourceAPI(c *client.Client, period time.Duration, servicesChan chan<- ServiceUpdate, endpointsChan chan<- EndpointsUpdate) {
+	servicesLW := cache.NewListWatchFromClient(c, "services", api.NamespaceAll, fields.Everything())
+	endpointsLW := cache.NewListWatchFromClient(c, "endpoints", api.NamespaceAll, fields.Everything())
 
-// NewSourceAPI creates a config source that watches for changes to the services and endpoints.
-func NewSourceAPI(servicesWatcher ServicesWatcher, endpointsWatcher EndpointsWatcher, period time.Duration, services chan<- ServiceUpdate, endpoints chan<- EndpointsUpdate) *SourceAPI {
-	config := &SourceAPI{
-		s: servicesReflector{
-			watcher:         servicesWatcher,
-			services:        services,
-			resourceVersion: "",
-			waitDuration:    period,
-			// prevent hot loops if the server starts to misbehave
-			reconnectDuration: time.Second * 1,
-		},
-		e: endpointsReflector{
-			watcher:         endpointsWatcher,
-			endpoints:       endpoints,
-			resourceVersion: "",
-			waitDuration:    period,
-			// prevent hot loops if the server starts to misbehave
-			reconnectDuration: time.Second * 1,
-		},
-	}
-	go util.Forever(func() { config.s.listAndWatch() }, period)
-	go util.Forever(func() { config.e.listAndWatch() }, period)
-	return config
-}
-
-func (r *servicesReflector) listAndWatch() {
-	r.run(&r.resourceVersion)
-	time.Sleep(wait.Jitter(r.reconnectDuration, 0.0))
-}
-
-func (r *endpointsReflector) listAndWatch() {
-	r.run(&r.resourceVersion)
-	time.Sleep(wait.Jitter(r.reconnectDuration, 0.0))
+	NewServicesSourceApiFromLW(servicesLW, period, servicesChan)
+	NewEndpointsSourceApiFromLW(endpointsLW, period, endpointsChan)
 }
 
-// run loops forever looking for changes to services.
-func (s *servicesReflector) run(resourceVersion *string) {
-	if len(*resourceVersion) == 0 {
-		services, err := s.watcher.List(labels.Everything())
-		if err != nil {
-			glog.Errorf("Unable to load services: %v", err)
-			// TODO: reconcile with pkg/client/cache which doesn't use reflector.
-			time.Sleep(wait.Jitter(s.waitDuration, 0.0))
-			return
+func NewServicesSourceApiFromLW(servicesLW cache.ListerWatcher, period time.Duration, servicesChan chan<- ServiceUpdate) {
```

Comment:
This doesn't really need to be public, and it's simple enough that it could be inlined. If you do want it to be public though, it should have a godoc comment.